From 637b591934be6bd327658ecf1c11c54eb77f531f Mon Sep 17 00:00:00 2001
From: Lu Yahan <yahan@iscas.ac.cn>
Date: Fri, 29 Jul 2022 09:52:02 +0800
Subject: [PATCH] Reland "[riscv][Cleanup] Use CmpInstanceTypeRange in
 MacroAssembler"

This is a reland of commit 859ff489615ad0aa62bcf5cfbf35ff6e4526897b

Original change's description:
> [riscv][Cleanup] Use CmpInstanceTypeRange in MacroAssembler
>
> Bug: v8:11325
>
> Change-Id: I2eae55b49ea01567460bd0adfbb819c893ce7cd7
> Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/3793210
> Auto-Submit: Yahan Lu <yahan@iscas.ac.cn>
> Reviewed-by: ji qiu <qiuji@iscas.ac.cn>
> Commit-Queue: ji qiu <qiuji@iscas.ac.cn>
> Cr-Commit-Position: refs/heads/main@{#82054}

Bug: v8:11325
Change-Id: I9db48ed2783a875b617d4161ce7405c0c32bebbe
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/3793466
Commit-Queue: ji qiu <qiuji@iscas.ac.cn>
Reviewed-by: ji qiu <qiuji@iscas.ac.cn>
Cr-Commit-Position: refs/heads/main@{#82063}
---
 src/codegen/riscv/macro-assembler-riscv.cc | 22 ++++++----------------
 1 file changed, 6 insertions(+), 16 deletions(-)

diff --git a/src/codegen/riscv/macro-assembler-riscv.cc b/src/codegen/riscv/macro-assembler-riscv.cc
index 26962c0f0f3..9fd9c7c4c4f 100644
--- a/src/codegen/riscv/macro-assembler-riscv.cc
+++ b/src/codegen/riscv/macro-assembler-riscv.cc
@@ -5624,22 +5624,12 @@ void MacroAssembler::AssertGeneratorObject(Register object) {
   Check(ne, AbortReason::kOperandIsASmiAndNotAGeneratorObject, kScratchReg,
         Operand(zero_reg));
 
-  GetObjectType(object, kScratchReg, kScratchReg);
-
-  Label done;
-
-  // Check if JSGeneratorObject
-  BranchShort(&done, eq, kScratchReg, Operand(JS_GENERATOR_OBJECT_TYPE));
-
-  // Check if JSAsyncFunctionObject (See MacroAssembler::CompareInstanceType)
-  BranchShort(&done, eq, kScratchReg, Operand(JS_ASYNC_FUNCTION_OBJECT_TYPE));
-
-  // Check if JSAsyncGeneratorObject
-  BranchShort(&done, eq, kScratchReg, Operand(JS_ASYNC_GENERATOR_OBJECT_TYPE));
-
-  Abort(AbortReason::kOperandIsNotAGeneratorObject);
-
-  bind(&done);
+  LoadMap(kScratchReg, object);
+  GetInstanceTypeRange(kScratchReg, kScratchReg, FIRST_JS_GENERATOR_OBJECT_TYPE,
+                       kScratchReg);
+  Check(
+      Uless_equal, AbortReason::kOperandIsNotAGeneratorObject, kScratchReg,
+      Operand(LAST_JS_GENERATOR_OBJECT_TYPE - FIRST_JS_GENERATOR_OBJECT_TYPE));
 }
 
 void MacroAssembler::AssertUndefinedOrAllocationSite(Register object,
-- 
2.35.1

